### PR TITLE
Feature/sync relations

### PR DESCRIPTION
**Pull request by zhex**
add 1-N, 1-1 relation sync to the DB.

### CODE REVIEW

**Comment by ianhe8x** [comment]
hold until open grant is accepted

**Comment by ianhe8x** [comment]
require changes on `@subql/cli` so codegen can recognize relations

**Comment by ianhe8x** [comment]
And we need new example to demonstrate the usage of these three type of relations, in the meantime, update docs.

**Comment by jiqiang90** [comment]
Closes #185

**Comment by sebastinez** [comment]
Hello folks,

Nice work on this implementation, really looking forward to it.
Do you already have any example how one should implement the mapping functions to make it work? Since the auto generated types with `subql codegen` it seems that some relationship is missing.

File: packages/node/src/indexer/store.service.ts [hunk 1]
```diff
@@ -14,19 +23,85 @@ export class StoreService {
 
   async syncSchema(
```
Comment (ianhe8x) [inline packages/node/src/indexer/store.service.ts:24]: store service don't need to know about graphql schema. I suggest you can make syncSchema() accept models and relations.

File: packages/node/src/indexer/store.service.ts [hunk 2]
```diff
@@ -14,19 +23,85 @@ export class StoreService {
...
     }
+
+    const extraQueries = [];
+
+    for (const relation of entitiesRelations.relations) {
+      const model = this.sequelize.model(relation.from);
+      const relatedModel = this.sequelize.model(relation.to);
+
+      if (relation.type === 'belongsTo') {
+        model.belongsTo(relatedModel, { foreignKey: `${relation.foreignKey}` });
```
Comment (ianhe8x) [inline packages/node/src/indexer/store.service.ts:51]: `` is not necessary

File: packages/node/src/indexer/store.service.ts [hunk 3]
```diff
@@ -14,19 +23,85 @@ export class StoreService {
...
       });
     }
+
+    const extraQueries = [];
+
+    for (const relation of entitiesRelations.relations) {
+      const model = this.sequelize.model(relation.from);
+      const relatedModel = this.sequelize.model(relation.to);
+
+      if (relation.type === 'belongsTo') {
```
Comment (ianhe8x) [inline packages/node/src/indexer/store.service.ts:50]:
```suggestion
switch case
```

File: packages/node/src/indexer/store.service.ts [hunk 4]
```diff
@@ -14,19 +23,85 @@ export class StoreService {
...
+        model.belongsTo(relatedModel, { foreignKey: `${relation.foreignKey}` });
+      } else if (relation.type === 'hasOne') {
+        const rel = model.hasOne(relatedModel, {
+          foreignKey: `${relation.foreignKey}`,
```
Comment (ianhe8x) [inline packages/node/src/indexer/store.service.ts:54]: `` is not necessary

File: packages/node/src/indexer/store.service.ts [hunk 5]
```diff
@@ -14,19 +23,85 @@ export class StoreService {
...
+        extraQueries.push(
+          commentConstraintQuery(
+            `${schema}.${rel.target.tableName}`,
+            fkConstraint,
+            tags,
+          ),
+          createUniqueIndexQuery(
+            schema,
+            relatedModel.tableName,
+            `${relation.foreignKey}`,
```
Comment (ianhe8x) [inline packages/node/src/indexer/store.service.ts:70]: `` is not necessary

File: packages/node/src/indexer/store.service.ts [hunk 6]
```diff
@@ -14,19 +23,85 @@ export class StoreService {
...
+          ),
+        );
+      } else if (relation.type === 'hasMany') {
+        const rel = model.hasMany(relatedModel, {
+          foreignKey: `${relation.foreignKey}`,
```
Comment (ianhe8x) [inline packages/node/src/indexer/store.service.ts:75]: `` is not necessary

File: packages/node/src/indexer/store.service.ts [hunk 7]
```diff
@@ -14,19 +23,85 @@ export class StoreService {
...
+        });
+        const fkConstraint = getFkConstraint(
+          rel.target.tableName,
+          rel.foreignKey,
+        );
+        const tags = smartTags({
+          fieldName: relation.foreignKey.slice(0, -2),
```
Comment (ianhe8x) [inline packages/node/src/indexer/store.service.ts:82]: will we have `id` as foreignKey for one-to-one case? I feel `slice()` is not safe

File: packages/node/src/indexer/store.service.ts [hunk 8]
```diff
@@ -14,19 +23,85 @@ export class StoreService {
...
+    }
+
     await this.sequelize.sync();
+    for (const query of extraQueries) {
+      await this.sequelize.query(query);
+    }
   }
 
   setTransaction(tx: Transaction) {
+    if (this.tx) {
```
Comment (ianhe8x) [inline packages/node/src/indexer/store.service.ts:102]: we have remove this code in the previous PR

**Review by ianhe8x — APPROVED** [review]

**Pull request merged**